File: scripts/import_local_images.py
#!/usr/bin/env python3
"""Import already-saved image files into images/shopify/ using manifest mappings.

Usage:
  python scripts/import_local_images.py /path/to/your/exported/images
"""

from __future__ import annotations

from pathlib import Path
import json
import shutil
import sys
from urllib.parse import urlparse

ROOT = Path(__file__).resolve().parents[1]
MANIFEST_PATH = ROOT / "images/shopify/manifest.json"
SUPPORTED_EXTS = {".jpg", ".jpeg", ".png", ".webp", ".gif", ".avif", ".jfif"}


def usage() -> int:
    print("Usage: python scripts/import_local_images.py /absolute/or/relative/path/to/local/images")
    return 1


def build_source_index(source_root: Path) -> dict[str, list[Path]]:
    index: dict[str, list[Path]] = {}
    for path in source_root.rglob("*"):
        if not path.is_file():
            continue
        if path.suffix.lower() not in SUPPORTED_EXTS:
            continue
        index.setdefault(path.name.lower(), []).append(path)
    return index


def source_filename_from_url(url: str) -> str:
    parsed = urlparse(url)
    return Path(parsed.path).name


def main(argv: list[str]) -> int:
    if len(argv) != 2:
        return usage()

    source_root = Path(argv[1]).expanduser().resolve()
    if not source_root.exists() or not source_root.is_dir():
        print(f"Source folder not found: {source_root}")
        return 1

    if not MANIFEST_PATH.exists():
        print(f"Manifest not found: {MANIFEST_PATH}")
        return 1

    entries = json.loads(MANIFEST_PATH.read_text())
    source_index = build_source_index(source_root)

    copied = 0
    missing = 0

    for entry in entries:
        local_rel = entry["local_path"]
        source_url = entry["source_url"]
        expected_name = source_filename_from_url(source_url).lower()

        candidates = source_index.get(expected_name, [])
        if not candidates:
            missing += 1
            continue

        src = candidates[0]
        dst = ROOT / local_rel
        dst.parent.mkdir(parents=True, exist_ok=True)
        shutil.copy2(src, dst)
        copied += 1

    print(f"Imported {copied} file mappings.")
    if missing:
        print(f"Missing {missing} mappings. Ensure the source folder contains the original Shopify filenames.")
        return 2

    print("Done. All mapped images were imported.")
    return 0


if __name__ == "__main__":
    raise SystemExit(main(sys.argv))

File: scripts/import-local-images.mjs
#!/usr/bin/env node
import fs from 'node:fs';
import path from 'node:path';
import { fileURLToPath } from 'node:url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
const root = path.resolve(__dirname, '..');
const manifestPath = path.join(root, 'images', 'shopify', 'manifest.json');
const supported = new Set(['.jpg', '.jpeg', '.png', '.webp', '.gif', '.avif', '.jfif']);

function usage() {
  console.log('Usage: node scripts/import-local-images.mjs /path/to/your/saved-images');
}

function walkFiles(dir, out = []) {
  for (const entry of fs.readdirSync(dir, { withFileTypes: true })) {
    const full = path.join(dir, entry.name);
    if (entry.isDirectory()) {
      walkFiles(full, out);
    } else if (entry.isFile()) {
      const ext = path.extname(entry.name).toLowerCase();
      if (supported.has(ext)) out.push(full);
    }
  }
  return out;
}

function sourceFileName(url) {
  const parsed = new URL(url);
  return path.basename(parsed.pathname).toLowerCase();
}

const sourceArg = process.argv[2];
if (!sourceArg) {
  usage();
  process.exit(1);
}

const sourceRoot = path.resolve(process.cwd(), sourceArg);
if (!fs.existsSync(sourceRoot) || !fs.statSync(sourceRoot).isDirectory()) {
  console.error(`Source folder not found: ${sourceRoot}`);
  process.exit(1);
}
if (!fs.existsSync(manifestPath)) {
  console.error(`Manifest not found: ${manifestPath}`);
  process.exit(1);
}

const manifest = JSON.parse(fs.readFileSync(manifestPath, 'utf8'));
const index = new Map();
for (const file of walkFiles(sourceRoot)) {
  const name = path.basename(file).toLowerCase();
  if (!index.has(name)) index.set(name, []);
  index.get(name).push(file);
}

let copied = 0;
let missing = 0;
for (const entry of manifest) {
  const expectedName = sourceFileName(entry.source_url);
  const candidates = index.get(expectedName) || [];
  if (candidates.length === 0) {
    missing += 1;
    continue;
  }

  const src = candidates[0];
  const dst = path.join(root, entry.local_path);
  fs.mkdirSync(path.dirname(dst), { recursive: true });
  fs.copyFileSync(src, dst);
  copied += 1;
}

console.log(`Imported ${copied} file mappings.`);
if (missing > 0) {
  console.log(`Missing ${missing} mappings. Ensure the source folder contains the original Shopify filenames.`);
  process.exit(2);
}
console.log('Done. All mapped images were imported.');
